File: military-FE/src/services/unitService.ts
import apiClient from '../utils/api'
import type { ApiResponse, Unit, CreateUnitRequest, UpdateUnitRequest, PaginationParams } from '../types/api'

export const unitService = {
  // Get all units
  getAll: async (params?: PaginationParams) => {
    const response = await apiClient.get<ApiResponse<Unit[]>>(
      '/units',
      { params },
    )
    return response.data
  },

  // Get unit by ID (code)
  getById: async (id: string) => {
    const response = await apiClient.get<ApiResponse<Unit>>(`/units/info/${id}`)
    return response.data
  },

  // Create unit
  create: async (data: CreateUnitRequest) => {
    const response = await apiClient.post<ApiResponse<Unit>>('/units', data)
    return response.data
  },

  // Update unit
  update: async (id: string, data: UpdateUnitRequest) => {
    const response = await apiClient.put<ApiResponse<Unit>>(`/units/${id}`, data)
    return response.data
  },

  // Delete unit
  delete: async (id: string) => {
    const response = await apiClient.delete<ApiResponse<{ message: string }>>(
      `/units/${id}`,
    )
    return response.data
  },
}

File: military_mobile/lib/services/unit_service.dart
import '../models/unit.dart';
import 'api_service.dart';

class UnitService {
  final ApiService _apiService = ApiService();

  List<Unit> _parseList(dynamic responseData) {
    final map = responseData as Map<String, dynamic>;
    final list = map['data'] as List<dynamic>? ?? [];
    return list.map((e) => Unit.fromJson(e as Map<String, dynamic>)).toList();
  }

  Unit _parseOne(dynamic responseData) {
    final map = responseData as Map<String, dynamic>;
    return Unit.fromJson(map['data'] as Map<String, dynamic>);
  }

  Future<List<Unit>> getAll() async {
    final response = await _apiService.get<Map<String, dynamic>>('/units');
    return _parseList(response.data);
  }

  Future<Unit> getById(String id) async {
    final response = await _apiService.get<Map<String, dynamic>>('/units/$id');
    return _parseOne(response.data);
  }

  Future<Unit> create({
    required String id,
    required String name,
    String? logo,
  }) async {
    final response = await _apiService.post<Map<String, dynamic>>(
      '/units',
      data: {
        'id': id,
        'name': name,
        'logo': logo,
      },
    );
    return _parseOne(response.data);
  }

  Future<Unit> update(
    String id, {
    required String name,
    String? logo,
  }) async {
    final response = await _apiService.put<Map<String, dynamic>>(
      '/units/$id',
      data: {
        'name': name,
        'logo': logo,
      },
    );
    return _parseOne(response.data);
  }

  Future<void> delete(String id) async {
    await _apiService.delete('/units/$id');
  }
}
